core/src/main/java/com/sequenceiq/cloudbreak/service/image/ImageCatalogService.java

```diff
@@ -288,6 +302,45 @@ public ImageCatalog update(ImageCatalog source) {
         return create(imageCatalog);
     }
 
+    public Images getApplicableImages(String stackName) throws CloudbreakImageCatalogException {
+        return getApplicableImages(CLOUDBREAK_DEFAULT_CATALOG_NAME, stackName);
+    }
+
+    public Images getApplicableImages(String imageCatalogName, String stackName) throws CloudbreakImageCatalogException {
+        IdentityUser user = authenticatedUserService.getCbUser();
+        Stack stack = stackService.getByName(stackName, user);
+
+        if (!Status.AVAILABLE.equals(stack.getStackStatus().getStatus()) || !Status.AVAILABLE.equals(stack.getCluster().getStatus())) {
+            throw new BadRequestException("To retrieve list of images for upgrade both stack and cluster have to be in AVAILABLE state");
+        }
+
+        String currentImageUuid = getCurrentImageUuid(stack);
+        StatedImages statedImages = getImages(imageCatalogName, stack.cloudPlatform());
+        List<Image> filteredBaseImages =
+                filterByApplicability(imageCatalogName, statedImages.getImageCatalogUrl(), stack, statedImages.getImages().getBaseImages(), currentImageUuid);
+        List<Image> filteredHdpImages =
+                filterByApplicability(imageCatalogName, statedImages.getImageCatalogUrl(), stack, statedImages.getImages().getHdpImages(), currentImageUuid);
+        List<Image> filteredHdfImages =
+                filterByApplicability(imageCatalogName, statedImages.getImageCatalogUrl(), stack, statedImages.getImages().getHdfImages(), currentImageUuid);
+        return new Images(filteredBaseImages, filteredHdpImages, filteredHdfImages, statedImages.getImages().getSuppertedVersions());
+    }
+
+    private String getCurrentImageUuid(Stack stack) {
+        try {
+            return componentConfigProvider.getImage(stack.getId()).getImageId();
+        } catch (CloudbreakImageNotFoundException e) {
+            LOGGER.warn("Could not find last used image when preparing upgrade image list for current cluster, exception: {}", e);
+            return "";
+        }
+    }
+
+    private List<Image> filterByApplicability(String imageCatalogName, String imageCatalogUrl, Stack stack, List<Image> imagesList, String currentUuid) {
+        return imagesList.stream()
+                .filter(x -> !x.getUuid().equals(currentUuid))
```

Comment:
maybe this "x" could be changed to image to improve readability, or at least to "i" to refer to image